modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php

```diff
@@ -289,88 +290,171 @@ public function buildPaneForm(array $pane_form, FormStateInterface $form_state,
+   *   The complete form structure.
+   */
+  protected function validateReturningCustomer(array &$pane_form, FormStateInterface $form_state, array &$complete_form) {
+    $values = $form_state->getValue($pane_form['#parents']);
+    $name_element = $pane_form['returning_customer']['name'];
+    $username = $values['returning_customer']['name'];
+    $password = trim($values['returning_customer']['password']);
+    // Generate the "reset password" url.
+    $query = !empty($username) ? ['name' => $username] : [];
+    $password_url = Url::fromRoute('user.pass', [], ['query' => $query])->toString();
+
+    if (empty($username) || empty($password)) {
+      $form_state->setError($pane_form['returning_customer'], $this->t('Unrecognized username or password. <a href=":url">Have you forgotten your password?</a>', [':url' => $password_url]));
+      return;
+    }
+    if (user_is_blocked($username)) {
+      $form_state->setError($name_element, $this->t('The username %name has not been activated or is blocked.', ['%name' => $username]));
+      return;
+    }
+    if (!$this->credentialsCheckFlood->isAllowedHost($this->clientIp)) {
+      $form_state->setErrorByName($name_element, $this->t('Too many failed login attempts from your IP address. This IP address is temporarily blocked. Try again later or <a href=":url">request a new password</a>.', [':url' => Url::fromRoute('user.pass')]));
+      $this->credentialsCheckFlood->register($this->clientIp, $username);
+      return;
+    }
+    elseif (!$this->credentialsCheckFlood->isAllowedAccount($this->clientIp, $username)) {
+      $form_state->setErrorByName($name_element, $this->t('Too many failed login attempts for this account. It is temporarily blocked. Try again later or <a href=":url">request a new password</a>.', [':url' => Url::fromRoute('user.pass')]));
+      $this->credentialsCheckFlood->register($this->clientIp, $username);
+      return;
+    }
+
+    $uid = $this->userAuth->authenticate($username, $password);
+    if (!$uid) {
+      $this->credentialsCheckFlood->register($this->clientIp, $username);
+      $form_state->setError($name_element, $this->t('Unrecognized username or password. <a href=":url">Have you forgotten your password?</a>', [':url' => $password_url]));
+    }
+    $form_state->set('logged_in_uid', $uid);
+  }
+
+  /**
+   * Validates the registration form and creates a new user.
+   *
+   * @param array $pane_form
+   *   The pane form, containing the following basic properties:
+   *   - #parents: Identifies the position of the pane form in the overall
+   *     parent form, and identifies the location where the field values are
+   *     placed within $form_state->getValues().
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The form state of the parent form.
+   * @param array $complete_form
+   *   The complete form structure.
+   */
+  protected function validateRegister(array &$pane_form, FormStateInterface $form_state, array &$complete_form) {
```

Comment:
It's odd that we have a validateRegister() method that does no validation, the entire logic is in the oddly named getAndValidateRegisterUserValues, which is only used in one place. What's the reasoning behind that split?